### ISSUE TITLE
about meanShape

### ISSUE THREAD

**Issue by zjjMaiMai** [MarekKowalski/DeepAlignmentNetwork#8]
i have some question about meanshape.

should i Normalize meanshape in BBox? 

I have not see you calc meanshape in your code.Can you share that? 

Thanks!

**Comment by MarekKowalski** [comment]
Hi,

I do not understand what you mean by: "should I normalize meanshape in BBox", can you give more details?

I have managed to find some functions which should get you an average face shape.
This is very old code from a time when I did not know anything about Python programming, so it is not very pretty ;)

Marek

```
def procrustes(shape, shape0):
    shapeVec = shape.flatten()
    shape0Vec = shape0.flatten()
 
    a = np.dot(shapeVec, shape0Vec) / np.linalg.norm(shapeVec)**2
    b = 0
    for i in range(shape0.shape[0]):
        b += shapeVec[2*i] * shape0Vec[2*i+1] - shapeVec[2*i+1] * shape0Vec[2*i] 
    b = b / np.linalg.norm(shapeVec)**2
     
    T = np.array([[a, b], [-b, a]])
     
    return np.dot(shapeVec.reshape((-1, 2)), T)
 
def alignShapes(points, refIndex=0, eps=10e-8, maxIter=50):
    points2 = points.copy()
    for i in range(points.shape[0]):
        points2[i] = points2[i] - np.mean(points2[i], axis=0)
     
    points2[refIndex] = points2[refIndex] #/ np.linalg.norm(points2[refIndex])
    x0 = points2[refIndex].copy()
    mean = points2[refIndex].copy()
     
    iteration = 1
    while True: 
        print(iteration)       
        for i in range(points2.shape[0]):
            points2[i] = procrustes(points2[i], mean)
         
        oldMean = mean.copy()
        mean = np.mean(points2, axis=0)
        mean = procrustes(mean, x0)
         
        if np.linalg.norm(oldMean - mean) < eps or iteration > maxIter:
            break
         
        iteration += 1
     
    mult = np.dot(mean.flatten(), mean.flatten())
    for i in range(points2.shape[0]):
        points2[i] = mult * (points2[i] / np.dot(points2[i].flatten(), mean.flatten()))
     
    return points2
```

**Comment by zjjMaiMai** [comment]
thank you for your answer And your code sincerely! 
Now I Re implementation your work on tensorflow. When i done it, I will OpenSource my project.

in your code , I see the LandmarkTranformLayer and AffineTransformLayer do similar transformation with last layer's output. Is that code run in GPU or CPU ?(sorry about this question,I'm not well with Lasagne|Theano )

And Do you remember the stage1's acc or err? I found that when i training stage1, the TrainSet's Acc is good,but the TestSet's ACC is low,maybe overfit . I wish to know that in your work to training stage1!

Thanks !

**Comment by MarekKowalski** [comment]
Hi,

Frankly I am not sure whether this is CPU or GPU either, I suspect CPU.

As shown in the article the mean error on the 300W private test set for a single stage is 5.02% of the distance between the outer eye corners. This is for the model trained on the smaller set (300W). I suggest you use this score as a reference.

Marek

**Comment by zjjMaiMai** [comment]
my reimplementation project: [https://github.com/zjjMaiMai/Deep-Alignment-Network-A-convolutional-neural-network-for-robust-face-alignment](url)

Thanks for your help!

**Comment by mariolew** [comment]
@MarekKowalski @zjjMaiMai I've implemented DAN using tensorflow based on your projects, and my implementation use the same data augmentation as this project.

**Comment by MarekKowalski** [comment]
I have added a link to your implementation to the Readme file.
